### DISCUSSION TITLE
Looks strange when use ast grep on language c mode

### DISCUSSION THREAD

**Discussion by lometsj** [ast-grep/ast-grep#1270] [Q&A] [answered]
# ⏯ Playground Link
[Playground link with relevant code](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImMiLCJxdWVyeSI6IiRDT1VOVCA9ICRcbiIsInJld3JpdGUiOiIiLCJjb25maWciOiJpZDogbWV0aG9kX3JlY2VpdmVyXG5ydWxlOlxuICBwYXR0ZXJuOiAkTSgkJCQpXG5cblxuIiwic291cmNlIjoidm9pZCB0ZXN0X2Z1bmMoKSB7XG4gICAgbWV0aG9kKCk7XG4gICAgb3RoZXJfbWV0aG9kKDEsIDIsIDMpO1xuICAgIG90aGVyX21ldGhvZCgxLCAyLCAzKTtcbn0ifQ==)

# 💻 Code

```c
#include <stdio.h>

#define test(x) (2*x)

int a = test(2);
int a[] = {1,2,3,4};

int main(){
        int b = test(2);
        int b[] = {1,2,3,4};
}
```

Pattern:

```yml
id: method_receiver
rule:
  pattern: test($$ARG)
```

# 🙁 Actual behavior

I try to find some macro call by using ast grep
However, when I use language c, it find nothing
After switch to cpp mode, it works as well

```
$ cat test.c
#include <stdio.h>

#define test(x) (2*x)

int a = test(2);
int a[] = {1,2,3,4};

int main(){
        int b = test(2);
        int b[] = {1,2,3,4};
        puts("a");
}
$ sg -p 'test($$$)' -l c test.c
$ sg -p 'test($$$)' -l cpp test.c
test.c
5│int a = test(2);
9│      int b = test(2);
$ sg --version
ast-grep 0.24.1
```

Why the result in c and cpp is different? And is it current for some reason?
I have confused about this for several days...

**Comment by HerringtonDarkholme** [comment, marked as answer]
You can use playground to debug your pattern.

<img width="520" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/fee2527a-a183-4b82-8b75-85ef06c36bd4">

`$M($$$A)` is parsed as macro. This is tree-sitter-c's behavior. It can be different from tree-sitter-cpp. The difference is caused by parser.


So you need [pattern object](https://ast-grep.github.io/guide/rule-config/atomic-rule.html#pattern-object) to make it a function call.

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoiYyIsInF1ZXJ5IjoiJE0oJCQkQSkiLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoiaWQ6IG1ldGhvZF9yZWNlaXZlclxucnVsZTpcbiAgcGF0dGVybjogJE0oJCQkKVxuXG5cbiIsInNvdXJjZSI6InZvaWQgdGVzdF9mdW5jKCkge1xuICAgIG1ldGhvZCgpO1xuICAgIG90aGVyX21ldGhvZCgxLCAyLCAzKTtcbiAgICBvdGhlcl9tZXRob2QoMSwgMiwgMyk7XG59In0=

**Reply by HerringtonDarkholme** [reply]
This rule will match funciton call. Note, `int x = test(2)` is parsed as function call.

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImMiLCJxdWVyeSI6InRlc3QoJCQkKSIsInJld3JpdGUiOiIiLCJjb25maWciOiJpZDogbWV0aG9kX3JlY2VpdmVyXG5ydWxlOlxuICBwYXR0ZXJuOiBcbiAgICBjb250ZXh0OiAkTSgkJCQpO1xuICAgIHNlbGVjdG9yOiBjYWxsX2V4cHJlc3Npb24iLCJzb3VyY2UiOiIjaW5jbHVkZSA8c3RkaW8uaD5cblxuI2RlZmluZSB0ZXN0KHgpICgyKngpXG5cbmludCBhID0gdGVzdCgyKTtcbmludCBhW10gPSB7MSwyLDMsNH07XG5cbmludCBtYWluKCl7XG4gICAgICAgIGludCBiID0gdGVzdCgyKTtcbiAgICAgICAgaW50IGJbXSA9IHsxLDIsMyw0fTtcbn0ifQ==

**Reply by lometsj** [reply]
Awesome, it is indeed the Tree-sitter that makes the difference
You let me know what I should do,very thanks

**Reply by HerringtonDarkholme** [reply]
I have added it in the example docs https://ast-grep.github.io/catalog/c/#match-function-call
